### PR TITLE
Order report

### PR DESCRIPTION
**Pull request by thasnim-cm**
- Order report
- Grid table

### CODE REVIEW

File: src/components/GridTable.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+          })
+        });
+        gridData.push(_data);
+      }
+      setRowDataList(gridData);
+    }
+  };
+
+  useEffect(() => {
+    generateGridData(props.columns);
```
Comment (raj-kumar-litmus) [inline src/components/GridTable.tsx]: dont pass `props.columns` as argument.
U can directly access props.columns inside `generateGridData`

File: src/components/GridTable.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      className={`grid-table cursor-pointer ${props.className}`}
+    >
+      {Array(props.columns).fill("").map((_, columnIndex) =>
+        <Column
+          key={`col${columnIndex + 1}`}
+          field={`data${columnIndex + 1}`}
+          body={(rowData) =>
+            rowData[columnIndex]?.[`data${columnIndex + 1}`] ?
+              <div
+                className={`flex justify-center items-center p-2 ${rowData[columnIndex].severity ? rowData[columnIndex].severity === SEVERITY.HIGH ? "text-yellow-200" : "text-green-400" : ""}`}
```
Comment (raj-kumar-litmus) [inline src/components/GridTable.tsx]: avoid nested ternary operators. Readability is difficult in such cases.
See if u can use  a state variable.

File: src/components/GridTable.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                onClick={() => props.onClick(rowData[columnIndex]?.[`data${columnIndex + 1}`])}>
+                {rowData[columnIndex].severity &&
```
Comment (raj-kumar-litmus) [inline src/components/GridTable.tsx]: null check

File: src/components/OrderSummaryCard.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import { OrderSummaryCardProps } from "../@types/components/commonTypes";
+
+const OrderSummaryCard: FC<OrderSummaryCardProps> = (props) => {
+  return (
+    <CustomCard className="summary-card my-1 text-gray-200 bg-black-200 w-full rounded-lg">
+      <div className="grid grid-cols-[repeat(auto-fill,minmax(91px,1fr))] gap-2">
+        {Object.entries(props.cardData)?.map(([key, value]: [string, Array<string | number> | string | number], i) => (
+          <div key={i}>
+            <div className="text-gray-400 text-10 min-h-[1.8rem]">{key}</div>
+            <div className="text-gray-300 text-base font-medium" title={value.toString()}>{value.length ? value[0] + " " + value[1] : value}</div>
```
Comment (raj-kumar-litmus) [inline src/components/OrderSummaryCard.tsx]: Please do null check for value

File: src/components/PieChart.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
+import type { ChartData, ChartOptions } from "chart.js";
+import { ArcElement, Chart as ChartJS, Legend, Tooltip } from "chart.js";
+import { FC } from "react";
+import { Pie } from "react-chartjs-2";
+import useScreenSize from "../hooks/useScreenSize";
+
+interface Props {
```
Comment (raj-kumar-litmus) [inline src/components/PieChart.tsx]: move this to commontypes.

File: src/components/PieChart.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        "rgba(153, 102, 255, 1)",
+        "rgba(255, 159, 64, 1)",
+      ],
+      borderWidth: 1,
+      lineTension: 0.8,
+    },
+  ],
+};
+
+const defaultOptions = {
```
Comment (raj-kumar-litmus) [inline src/components/PieChart.tsx]: move this to chartConfig.tsx

File: src/components/PieChart.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  height?: string;
+  width?: string;
+  xsHeight?: string;
+  xsWidth?: string;
+}
+
+ChartJS.register(ArcElement, Tooltip, Legend);
+
+const defaultData = {
+  labels: ["Red", "Blue", "Yellow", "Green", "Purple", "Orange"],
```
Comment (raj-kumar-litmus) [inline src/components/PieChart.tsx]: move this to chartconfig.tsx

File: src/views/orderReport.tsx [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+import LineChart from "../components/LineChart";
+import HorizontalTable from "../components/HorizontalTable";
+import CustomTab from "../components/common/customtab";
+import CustomTable from "../components/common/customtable";
+import Loader from "../components/loader";
+import OrderSummaryCard from "../components/OrderSummaryCard";
+import PieChart from "../components/PieChart";
+import { HOURLY_ORDER_TREND_LINE_CHART_OPTION, ORDER_REPORT_PIE_CHART_OPTIONS } from "../config/chartConfig";
+import { URL_HOURLYREPORT } from "../constants/apiConstants";
+import { CHANNELWISE_DATA_LABELS, HOURLY_TREND_DATA_LABELS, LOCALEWISE_DATA_LABELS, ORDER_PIE_CHART_LABELS, ORDER_REPORT_CHART_LABELS, ORDER_REPORT_LABELS, ORDER_REPORT_THRESHOLDS, ORDER_SUMMARY_LABELS, PAYMENTWISE_DATA_LABELS, REPORT_TABS, SCREEN_WIDTH } from "../constants/appConstants";
```
Comment (raj-kumar-litmus) [inline src/views/orderReport.tsx]: This line should be in multiple lines if prettier is working as expected.
Can you check ?

File: src/views/orderReport.tsx [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+    } catch (error) {
+      console.error("Error while fetching data:", error);
+    } finally {
+      setIsLoading(false);
+    }
+  };
+
+  const getOrderSummary = () => {
+    let data = hourlyReport.summary;
+    let summary = [
```
Comment (raj-kumar-litmus) [inline src/views/orderReport.tsx]: const

File: src/views/orderReport.tsx [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+        [ORDER_SUMMARY_LABELS.US_AVERAGE_ORDER_VALUE]: [data.usAverageOrderValue, ORDER_REPORT_LABELS.USD],
+        [ORDER_SUMMARY_LABELS.CA_AVERAGE_ORDER_VALUE]: [data.caAverageOrderValue, ORDER_REPORT_LABELS.CAD]
+      },
+    ]
+    setOrderSummary(summary);
+  }
+
+  const getChannelWiseOrder = () => {
+    let data = hourlyReport.channelWiseOrderReport;
+    let datasets = [
```
Comment (raj-kumar-litmus) [inline src/views/orderReport.tsx]: always const over let.
Check in all places.

File: src/views/orderReport.tsx [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+      label: HOURLY_TREND_DATA_LABELS.LAST_MONTH,
+      data: data.lastMonth,
+      borderColor: 'rgb(255, 69, 96)'
+    };
+    let lastYearData = {
+      label: HOURLY_TREND_DATA_LABELS.LAST_YEAR,
+      data: data.lastYear,
+      borderColor: 'rgb(119, 93, 208)'
+    };
+
```
Comment (raj-kumar-litmus) [inline src/views/orderReport.tsx:210]: Try this approach to reduce number of lines. 
In the current format, the code looks repetitive.

**Untested code. Please test the same.**

```
[
  [HOURLY_TREND_DATA_LABELS.TODAY,data.today,'rgb(0, 143, 251)'],
  [HOURLY_TREND_DATA_LABELS.YESTERDAY,data.yesterday,'rgb(0, 227, 150)'],
  [HOURLY_TREND_DATA_LABELS.LAST_WEEK,data.lastWeek,'rgb(254, 176, 25)'],
]
.map(({label, data, borderColor}) => ({
  label,
  data,
  borderColor,
}))
```

File: src/views/orderReport.tsx [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+    let historicalOrderData = hourlyReport.hourlyHistoricalOrderData;
+    let summary = hourlyReport.summary;
+    let time = ORDER_REPORT_CHART_LABELS.HOURLY_TIMESTAMPS.slice(1);
+    let totalData = [getTotalData(historicalOrderData.lastYear),
+    getTotalData(historicalOrderData.lastMonth), getTotalData(historicalOrderData.lastWeek),
+    getTotalData(historicalOrderData.yesterday), summary.totalOrdersPlaced,
+    summary.usOrdersPlaced, summary.caOrdersPlaced,
+    summary.usSales, summary.caSales, summary.totalSales]
+    setLocaleWiseTotalData(totalData);
+    let orderData = time.map((t, i) => {
```
Comment (raj-kumar-litmus) [inline src/views/orderReport.tsx]: single line method. no  need to `return` directly pass it to the arrow function.

File: src/views/orderReport.tsx [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+      }
+    });
+    setPaymentWiseHourlyOrderData(orderData);
+  }
+
+  const checkTableData = (data, time) => {
+    return ((!data && (Number(CURRENT_PST_DATE.getHours()) < Number(time.split(":")[0])))) ? "" : data;
+  }
+
+  const getTotalData = (data: number[]) => {
```
Comment (raj-kumar-litmus) [inline src/views/orderReport.tsx]: use `reduce`

File: src/views/orderReport.tsx [hunk 14]
```diff
@@ -0,0 +1,467 @@
...
+    };
+    setHourlyOrderTrendData(chartData);
+  }
+
+  const getLocaleWiseHourlyOrderData = () => {
+    let slicedOrderData = hourlyReport.hourlySlicedOrderData;
+    let historicalOrderData = hourlyReport.hourlyHistoricalOrderData;
+    let summary = hourlyReport.summary;
+    let time = ORDER_REPORT_CHART_LABELS.HOURLY_TIMESTAMPS.slice(1);
+    let totalData = [getTotalData(historicalOrderData.lastYear),
```
Comment (raj-kumar-litmus) [inline src/views/orderReport.tsx]: always declare a method first before invoking it.
Here `getTotalData` is being declared after this method.

**Comment by raj-kumar-litmus** [comment]
1. Always const over let.
2. Check prettier configuration. Lots of long lines are not wrapped into multiple lines.

File: src/index.css [hunk 15]
```diff
@@ -995,12 +998,72 @@ hr.horizontalLine.blue {
...
 }
-.quick-link-arrow img{
+.quick-link-arrow img {
   @apply h-7p w-7p;
 }
+.report-tab.p-tabview {
+  width: 335px;
+}
+.p-datatable .p-datatable-tfoot > tr > td {
+  @apply !bg-[#22262C] text-[#FFFFFF] border-0 border-[#292E36] p-3;
```
Comment (sarath-sasi-panickaveetil) [inline src/index.css]: Use the existing tailwind class (incase any) or extend it in the tailwind config.

File: src/index.css [hunk 16]
```diff
@@ -995,12 +998,72 @@ hr.horizontalLine.blue {
...
+  @apply rounded-bl-lg;
+}
+.report-table .p-datatable-tfoot tr:last-child td:last-child {
+  @apply rounded-br-lg;
+}
+.channel-table .p-datatable-tbody tr td {
+  @apply max-w-[82px] min-w-[82px];
+}
+.payment-table .p-datatable-thead tr th {
+  @apply py-[.75rem] px-[1.7px];
```
Comment (sarath-sasi-panickaveetil) [inline src/index.css]: py-3 can be used instead of py-[.75rem]. Please change in below class also

File: src/components/GridTable.tsx (hunk 1 above)
Comment (thasnim-cm) [reply src/components/GridTable.tsx]: Done

File: src/views/orderReport.tsx (hunk 8 above)
Comment (thasnim-cm) [reply src/views/orderReport.tsx]: I had already compared my prettier config with Adithya and Sarath. We couldn't find any difference. This formatting is not working in my vs code

File: src/views/orderReport.tsx (hunk 9 above)
Comment (thasnim-cm) [reply src/views/orderReport.tsx]: done

File: src/views/orderReport.tsx (hunk 10 above)
Comment (thasnim-cm) [reply src/views/orderReport.tsx]: Added const instead of let where the variables are not reassigned

File: src/views/orderReport.tsx (hunk 11 above)
Comment (thasnim-cm) [reply src/views/orderReport.tsx:210]: done

File: src/views/orderReport.tsx (hunk 12 above)
Comment (thasnim-cm) [reply src/views/orderReport.tsx]: done

File: src/views/orderReport.tsx (hunk 13 above)
Comment (thasnim-cm) [reply src/views/orderReport.tsx]: done

File: src/components/GridTable.tsx [hunk 17]
```diff
@@ -10,14 +10,19 @@ import { GridTableProps } from "../@types/components/commonTypes";
 const GridTable: FC<GridTableProps> = (props) => {
   const [rowDataList, setRowDataList] = useState<any[]>([]);
 
-  const generateGridData = (columns) => {
+  const getSeverityStyles = (data) => {
```
Comment (raj-kumar-litmus) [inline src/components/GridTable.tsx:13]: single line arrow function.

File: src/components/SectionTitle.tsx [hunk 18]
```diff
@@ -0,0 +1,10 @@
+import { FC } from "react";
+import { SectionTitleProps } from "../@types/components/commonTypes";
+
+const SectionTitle: FC<SectionTitleProps> = (props) => {
```
Comment (raj-kumar-litmus) [inline src/components/SectionTitle.tsx:4]: single line arrow function.

**Pull request merged**